net, rpc: expose high bandwidth mode state via getpeerinfo

theStack opened the pull request:
Fixes #19676, "_For every peer expose through getpeerinfo RPC whether or not we selected them as HB peers, and whether or not they selected us as HB peers._" See [BIP152](https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki), in particular the [protocol flow diagram](https://github.com/bitcoin/bips/raw/master/bip-0152/protocol-flow.png).  The newly introduced states are changed on the following places in the code:
* on reception of a `SENDCMPCT` message with valid version, the field `m_highbandwidth_from` is changed depending on the first integer parameter in the message (1=high bandwidth, 0=low bandwidth), i.e. it just mirrors the field `CNodeState.fPreferHeaderAndIDs`.
* after adding a `SENDCMPCT` message to the send queue, the field `m_highbandwidth_to` is changed depending on how the first integer parameter is set (same as above)

Note that after receiving `VERACK`, the node also sends `SENDCMPCT`, but that is only to announce the preferred version and never selects high-bandwidth mode, hence there is no need to change the state variables there, which are initialized to `false` anyways.

theStack committed:
net: save high-bandwidth mode states in CNodeStats

theStack committed:
rpc: expose high-bandwidth mode states via getpeerinfo

theStack committed:
doc: release note for new getpeerinfo fields "bip152_hb_{from,to}"

Co-authored-by: Jon Atack <jon@atack.com>

theStack committed:
test: add test for high-bandwidth mode states in getpeerinfo

jonatack commented:
> Would be nice to have `bitcoin-cli -netinfo t` expose this info in the future :)

Yes. Along with other new ones like `last_transaction`, `last_block`, `conn_type`.

jonatack commented on src/net_processing.cpp:
```diff
@@ -2629,8 +2631,10 @@ void ProcessMessage(
-            if (State(pfrom.GetId())->fWantsCmpctWitness == (nCMPCTBLOCKVersion == 2)) // ignore later version announces
+            if (State(pfrom.GetId())->fWantsCmpctWitness == (nCMPCTBLOCKVersion == 2)) { // ignore later version announces
                 State(pfrom.GetId())->fPreferHeaderAndIDs = fAnnounceUsingCMPCTBLOCK;
+                pfrom.m_highbandwidth_from = fAnnounceUsingCMPCTBLOCK;
```
f9cc66bc An explanatory comment here may be nice; I found this the least clear line of this PR changeset.

theStack replied:
Done.

jonatack commented on test/functional/p2p_compactblocks.py:
```diff
@@ -785,6 +785,40 @@ def announce_cmpct_block(node, peer):
+
+        # Asserts that the RPC getpeerinfo boolean result fields "highbandwidth_{from,to}"
+        # for the last peer of a given node match the given parameters
+        def assert_highbandwidth_states(node, hb_from, hb_to):
```
edb4ad4 suggestions
```diff
-        # Asserts that the RPC getpeerinfo boolean result fields "highbandwidth_{from,to}"
-        # for the last peer of a given node match the given parameters
-        def assert_highbandwidth_states(node, hb_from, hb_to):
+        # Assert the getpeerinfo `highbandwidth_to/highbandwith_from` boolean
+        # fields for the last peer of a given node match the given parameters.
+        def assert_highbandwidth_states_of_last_peer(node, hb_from, hb_to):
```

theStack replied:
Shortened the comment a bit and also reordered the parameters. Didn't want to change the function name to be even longer though (considering it's so close to where its used, possibly even closer after #19781 would be merged) everything should be clear.

jonatack reviewed (commented):
Tested and globally looks good. A few suggestions follow.

sipa commented:
Concept ACK.

I'd suggest putting "bip152" somewhere in the name of the field, as the concept of high bandwidth mode is very specific to that BIP (and really doesn't dramatically increase actual bandwidth, as an unqualified name may make seem).

theStack commented:
@sipa: Agreed that referring to the BIP in the name would make sense here.
@jonatack: Thanks for the detailled review! I plan to address all of your suggestions. 

Before I take any further changes, I'd like to discuss the naming. Some suggestions:
- `bip152_highbandwidth_{from,to}`
- `bip152_hb_{from,to}` (shorter, but not obvious)
- `bip152_cmpctblk_announce_{from,to}` (avoids the "high-bandwidth" terminology)
- ...

Does anyone have preferrences here or any other idea? My favourite would simply be `bip152_highbandwidth_{from,to}` (though being quite long).

jonatack commented:
> My favourite would simply be `bip152_highbandwidth_{from,to}`

I agree with you. Might be good to have input from @TheBlueMatt or @gmaxwell.

instagibbs commented:
> bip152_highbandwidth_{from,to}
> bip152_hb_{from,to} (shorter, but not obvious)

These seems fine. Anyone who is using this stuff should be familiar with the BIP and if not read it anyways.

naumenkogs commented:
Concept ACK.

theStack commented:
Rebased on master.
Will still wait a bit for more naming opinions before tackling review suggestions. (One could argue that the rebase was pointless then, but I think PRs having the "needs rebase" label are perceived as less attractive for review and more easily ignored.)

Currently `bip152_bandwidth_{from,to}` (preferred by 2, jonatack and my humble self) is favoured over `bip152_bh_{from,to}` (preferred by 1, instagibbs).

theStack commented on doc/release-notes-19776.md:
```diff
@@ -0,0 +1,9 @@
+  blocks high-bandwidth peer (`..._from`) or whether we selected a peer to be in
+  compact blocks high-bandwidth mode (`..._to`), respectively. High-bandwidth
+  peers are supposed to send new block announcements via a `cmptblock` message,
+  rather than the usual inv/headers announcements. See BIP 152 for more details.
```
Thanks, I like that much better! Adapted it, removed the indentation (as other PR review notes currently in master also don't have it) and added you as co-author.

theStack commented on src/rpc/net.cpp:
```diff
@@ -110,6 +110,8 @@ static UniValue getpeerinfo(const JSONRPCRequest& request)
                             {RPCResult::Type::STR, "subver", "The string version"},
                             {RPCResult::Type::BOOL, "inbound", "Inbound (true) or Outbound (false)"},
+                            {RPCResult::Type::BOOL, "highbandwidth_from", "Whether peer selected us as (compact blocks) high-bandwidth peer"},
+                            {RPCResult::Type::BOOL, "highbandwidth_to", "Whether we selected peer as (compact blocks) high-bandwidth peer"},
```
Good point, done.

theStack commented on test/functional/p2p_compactblocks.py:
```diff
@@ -785,6 +785,40 @@ def announce_cmpct_block(node, peer):
+            assert_equal(peerinfo['highbandwidth_from'], hb_from)
+            assert_equal(peerinfo['highbandwidth_to'], hb_to)
+
+        # initially, both nodes haven't selected the other peer as high-bandwidth yet
```
Thanks, fixed.

theStack commented:
Decided for the name `bip152_highbandwidth_{to,from}` for now (both for the exposed RPC result fields and the internal variable names) and force-pushed with most of the suggestions from jonatack.

gmaxwell commented:
The patch I carry locally for this calls it 

+            obj.pushKV("bip152_hb", 

but I don't think it matter other than it should mention BIP152.

jonatack commented on src/rpc/net.cpp:
```diff
@@ -110,6 +110,8 @@ static UniValue getpeerinfo(const JSONRPCRequest& request)
                             {RPCResult::Type::STR, "subver", "The string version"},
                             {RPCResult::Type::BOOL, "inbound", "Inbound (true) or Outbound (false)"},
+                            {RPCResult::Type::BOOL, "bip152_highbandwidth_to", "Whether we selected peer as (compact blocks) high-bandwidth peer"},
+                            {RPCResult::Type::BOOL, "bip152_highbandwidth_from", "Whether peer selected us as (compact blocks) high-bandwidth peer"},
```
The names have indeed become very long, and they should arguably be written as `bip152_high_bandwidth_{to/from}`, so I'd now vote for `bip152_hb_{to/from}`.

```
    "version" : n,                               (numeric) The peer version, such as 70001
    "subver" : "str",                            (string) The string version
    "inbound" : true|false,                      (boolean) Inbound (true) or Outbound (false)
    "bip152_highbandwidth_to" : true|false,      (boolean) Whether we selected peer as (compact blocks) high-bandwidth peer
    "bip152_highbandwidth_from" : true|false,    (boolean) Whether peer selected us as (compact blocks) high-bandwidth peer
    "addnode" : true|false,                      (boolean) Whether connection was due to addnode/-connect or if it was an automatic/inbound connection
    "startingheight" : n,                        (numeric) The starting height (block) of the peer
```

theStack replied:
Done, see comment below.

jonatack commented on doc/release-notes-19776.md:
```diff
@@ -0,0 +1,9 @@
+Updated RPCs
+------------
+
+The `getpeerinfo` RPC returns two new boolean fields, `bip152_highbandwidth_to`
```
```suggestion
- The `getpeerinfo` RPC returns two new boolean fields, `bip152_highbandwidth_to`
```

theStack replied:
Done.

jonatack reviewed (commented):
ACK a286f89a9c0

Happy to re-review if you opt for the shorter naming.

theStack commented:
Now that three people voted for the short version (instagibbs, jonatack, and indirectly also gmaxwell), I changed the RPC result fields to `bip152_hb_{from,to}`. The internal state variable names in `CNode{Stats}` are still kept in the long version though.

jonatack reviewed (commented):
Code review ACK bff461c per `git diff a286f89 bff461c`

theStack commented:
Rebased on master and shortened the test by using the parameterized ctor for `msg_sendcmpct`, now that #19781 has been merged.

jonatack commented:
Code review re-ACK bc8cc02 per `git range-diff 831b0ec bff461c bc8cc02`

naumenkogs commented:
Code review ACK bc8cc02

theStack commented:
Rebased on master (was needed after the recent merge of #17785). Thanks a lot for reviewing, jonatack and naumenkogs!

naumenkogs commented:
reACK 4df1d1244cd334fdb946de698fe39eba1b11147b

jonatack commented:
Code review re-ACK 4df1d1244cd334fdb946de698fe39eba1b11147b per ` git range-diff 7737603 bc8cc02 4df1d12`

sipa commented on src/net.h:
```diff
@@ -934,6 +936,10 @@ class CNode
     uint256 hashContinue;
     std::atomic<int> nStartingHeight{-1};
+    // We selected peer as (compact blocks) high-bandwidth peer (BIP152)
+    bool m_bip152_highbandwidth_to{false};
```
As these are mutable fields, add a `GUARDED_BY` annotation?

theStack replied:
Good point, I'm not sure though on what to put into the annotation, i.e. guarded by _what_? I see three possibilities here to protecting concurrent access to the boolean fields:
- introducing a new Mutex (seems overkill)
- guard with cs_main, as that Mutex is already guaranteed to be held when the flags are written (seems overkill as well, considering this is not validation-critical at all and just about some statistics; locking cs_main in CNode::copyStats() just feels wrong)
- simply using an atomic bool

Any thoughts about this?

sipa replied:
Atomic bools seem easiest.

theStack commented:
Rebased on master, changed the data type of the introduced high-bandwidth flags `CNode::m_bip152_highbandwidth_{to,from}` to atomic bools (see discussion https://github.com/bitcoin/bitcoin/pull/19776#discussion_r495482994) and fixed typos in the comments (s/sendcmpt/sendcmpct/).

naumenkogs commented:
reACK 343dc4760fd2407895fc8b3417a504b194429156

jonatack commented:
re-ACK 343dc4760fd2407895fc8b3417a504b194429156 per `git range-diff 7ea6499 4df1d12 343dc47`

theStack commented:
Since there hasn't been any activity for more than three weeks: is there anything else I can do for this PR? It has two ACKs (jonatack, naumenkogs) and two Concept ACKs (sipa, practicalswift), I think it should be ready for merge.

jonatack commented:
I reckon it's not a reflection on this PR (which seems good AFAICS), just that attention has been focused on reviewing key features before the feature freeze October 15 and now on fixing bugs before branch-off.

MarcoFalke commented:
Looks like this has missed the feature freeze. Can be merged after Nov 1st
